### PR TITLE
Update navigation API article

### PR DESCRIPTION
**Pull request by jakearchibald**
Preview: https://deploy-preview-3390--developer-chrome-com.netlify.app/docs/web-platform/navigation-api/

- Updates the documentation and code examples for the switch from `transitionWhile` to `intercept`
- Documents scroll handling
- Documents focus handling
- More detail on the types of navigations you do/don't want to intercept
- Less focus on `updateCurrentEntry`

@domenic: mind reviewing?

@samthor: I don't think there's anything here you'd disagree with, but your name is still on this, so you might want to take a quick look.

### CODE REVIEW

**Review by samthor — COMMENTED** [review]
Looks sane to me.

File: site/en/docs/web-platform/navigation-api/index.md [hunk 1]
```diff
@@ -39,32 +40,47 @@ In most cases, it lets your code override the browser's default behavior for tha
...
 ```js
 navigation.addEventListener('navigate', navigateEvent => {
-  switch (navigateEvent.destination.url) {
-    case 'https://example.com/':
-      navigateEvent.transitionWhile(loadIndexPage());
-      break;
-    case 'https://example.com/cats':
-      navigateEvent.transitionWhile(loadCatsPage());
-      break;
+  const url = new URL(navigateEvent.destination.url);
```
Comment (domenic) [inline site/en/docs/web-platform/navigation-api/index.md:47]: Maybe move this below the early bail-out.

File: site/en/docs/web-platform/navigation-api/index.md [hunk 2]
```diff
@@ -39,32 +40,47 @@ In most cases, it lets your code override the browser's default behavior for tha
...
+
+  if (url.pathname === '/') {
+    navigateEvent.intercept({handler: loadIndexPage});
+  } else if (url.pathname === '/cats/') {
+    navigateEvent.intercept({handler: loadCatsPage});
   }
 });
 ```
 
 You can intercept the navigation in one of two ways:
```
Comment (domenic) [inline site/en/docs/web-platform/navigation-api/index.md:61]:
```suggestion
You can deal with the navigation in one of two ways:
```

or similar?

File: site/en/docs/web-platform/navigation-api/index.md [hunk 3]
```diff
@@ -39,32 +40,47 @@ In most cases, it lets your code override the browser's default behavior for tha
...
 
-This example calls `transitionWhile()` on the event with a promise generated from async functions.
+{% Aside %}
+An earlier version of this API used `navigateEvent.transitionWhile(promise)` rather than `navigateEvent.intercept({ handler })`.
+
+`intercept` landed in Chrome 105.
+`transitionWhile` is deprecated, and will be removed in Chrome 108.
+{% endAside %}
+
+This example calls `intecept()` on the event with a promise generated from an async function.
```
Comment (domenic) [inline site/en/docs/web-platform/navigation-api/index.md:73]: Doesn't seem like a correct sentence anymore.

File: site/en/docs/web-platform/navigation-api/index.md [hunk 4]
```diff
@@ -92,90 +108,252 @@ Links might come and go on your page, and they're not the only way users can nav
 E.g., they may submit a form or even use an [image map].
 Your page might deal with these, but there's a long tail of possibilities which could just be simplified—something that the new Navigation API achieves.
 
+Additionally, the above doesn't handle back/forward navigation. There's another event for that, `"popstate"`.
```
Comment (domenic) [inline site/en/docs/web-platform/navigation-api/index.md:111]: Events in this article seem to be without code font (e.g. lots of instances of "navigate", not `"navigate"`).

File: site/en/docs/web-platform/navigation-api/index.md [hunk 5]
```diff
@@ -92,90 +108,252 @@ Links might come and go on your page, and they're not the only way users can nav
...
 
-As such, this API introduces a semantic concept that the browser understands: an SPA navigation is currently occurring, over time, changing the document from a previous URL and state to a new one.
-This has a number of potential benefits, including accessibility: browsers can surface the beginning, end, or potential failure of a navigation.
-Chrome, for example, activates its native loading indicator, and allows the user to interact with the stop button. (This doesn't currently happen when the user navigates via the back/forward buttons, but that [will be fixed soon][loading-crbug].)
+The key properties are:
 
-### Transition Success and Failure
+`canIntercept`
+: If this is false, you can't intercept the navigation.
+Cross-origin navigations, and cross-document traversals cannot be intercepted.
```
Comment (domenic) [inline site/en/docs/web-platform/navigation-api/index.md:125]:
```suggestion
Cross-origin navigations and cross-document traversals cannot be intercepted.
```

File: site/en/docs/web-platform/navigation-api/index.md [hunk 6]
```diff
@@ -92,90 +108,252 @@ Links might come and go on your page, and they're not the only way users can nav
...
+`downloadRequest`
+: If this is true, the navigation was initiated by a link with a `download` attribute.
+In most cases, you don't need to intercept this.
 
-These issues are being [discussed on GitHub](https://github.com/WICG/navigation-api/issues/66).
-{% endAside %}
+`formData`
+: If this isn't null, then this navigation is part of a POST form submission.
+Make sure you take this into account when handling the navigation.
+If you only want to handle GET navigations, avoid intercepting navigations where `formData` is not null.
```
Comment (domenic) [inline site/en/docs/web-platform/navigation-api/index.md:141]: Maybe mention that actually intercepting this is pretty common, to progressively enhance `<form>` submissions into Ajax? Forward-link to the `#form-submissions` heading?

File: site/en/docs/web-platform/navigation-api/index.md [hunk 7]
```diff
@@ -92,90 +108,252 @@ Links might come and go on your page, and they're not the only way users can nav
...
+
 Personally, the History API often _feels_ like it could go some way to help with these possibilities.
 However, it really only has two surface areas: responding if the user presses Back or Forward in their browser, plus pushing and replacing URLs.
 It doesn't have an analogy to "navigate", except if you manually set up listeners for, e.g., click events, as demonstrated above.
 
-## Transition
+## Deciding how to handle a navigation
 
-When your code calls `transitionWhile()` from within its "navigate" listener, it informs the browser that it's now preparing the page for the new, updated state; and that the navigation may take some time. The `Promise` you pass to `transitionWhile()` tells the browser how long the navigation takes.
+The `navigateEvent` contains a lot of information about the navigation that you can use to decide whether to intercept a particular navigation.
```
Comment (domenic) [inline site/en/docs/web-platform/navigation-api/index.md:119]:
```suggestion
The `navigateEvent` contains a lot of information about the navigation that you can use to decide how to deal with a particular navigation.
```

maybe? Since this is also about whether you want to cancel.

File: site/en/docs/web-platform/navigation-api/index.md [hunk 8]
```diff
@@ -92,90 +108,252 @@ Links might come and go on your page, and they're not the only way users can nav
...
+    // If this is a form submission,
+    // let that go to the server.
+    navigationEvent.formData
+  );
+}
+```
+
+## Intercepting
+
+When your code calls `intercept({ handler })` from within its "navigate" listener, it informs the browser that it's now preparing the page for the new, updated state; and that the navigation may take some time.
```
Comment (domenic) [inline site/en/docs/web-platform/navigation-api/index.md:168]:
```suggestion
When your code calls `intercept({ handler })` from within its "navigate" listener, it informs the browser that it's now preparing the page for the new, updated state, and that the navigation may take some time.
```

File: site/en/docs/web-platform/navigation-api/index.md [hunk 9]
```diff
@@ -92,90 +108,252 @@ Links might come and go on your page, and they're not the only way users can nav
...
+    });
+  }
 });
 ```
 
-The "navigateerror" event listener, which receives an `ErrorEvent`, is particularly handy as it's guaranteed to receive any errors from your code that's setting up a new page.
-You can simply `await fetch()` knowing that if the network is unavailable, the error will eventually be routed to `"navigateerror"`.
+This not only avoids URL resolution issues, it also feels fast because you're instantly responding to the user.
 
 ### Abort Signals
```
Comment (domenic) [inline site/en/docs/web-platform/navigation-api/index.md:222]:
```suggestion
### Abort signals
```

File: site/en/docs/web-platform/navigation-api/index.md [hunk 10]
```diff
@@ -92,90 +108,252 @@ Links might come and go on your page, and they're not the only way users can nav
...
+      },
+    });
   }
 });
 ```
 
-## Navigation Entries
+### Scroll handling
+
+By default, the browser will attempt to handle scrolling automatically.
```
Comment (domenic) [inline site/en/docs/web-platform/navigation-api/index.md:269]: Maybe mention that this only happens if you call `intercept()`? Otherwise you get today's current behavior with no delays and such.

File: site/en/docs/web-platform/navigation-api/index.md [hunk 11]
```diff
@@ -217,28 +395,72 @@ console.log(navigation.currentEntry.getState());
...
+Then, other code can hear about this change via the `"currententrychange"` event:
+
+```js
+navigation.addEventListener('currententrychange', () => {
+  console.log(navigation.currentEntry.getState());
+});
+```
+
+However, this splits your state-handing code between the `"navigate"` event and the `"currententrychange"` event.
+It's generally better to handle state in one place, the `"navigate"` event.
```
Comment (domenic) [inline site/en/docs/web-platform/navigation-api/index.md:452]: I appreciate the deemphasis on updateCurrentEntry(). However I hesitate a bit to leave this section without mentioning the react-to vs. initiate distinction that https://github.com/wicg/navigation-api#setting-the-current-entrys-state-without-navigating mentions. I'm not sure how to do that without expanding the section and thus perhaps overemphasizing updateCurrentEntry()... maybe just an explainer link?

File: site/en/docs/web-platform/navigation-api/index.md [hunk 12]
```diff
@@ -294,7 +516,7 @@ In fact, it will always be `undefined` in those cases.
     height="320",
     muted="true" %}
   <figcaption>
-    <a href="https://wiry-tricolor-lipstick.glitch.me" target="_blank">Demo of opening from left or right</a>
+    <a href="https://gigantic-honored-octagon.glitch.me/" target="_blank">Demo of opening from left or right</a>
```
Comment (domenic) [inline site/en/docs/web-platform/navigation-api/index.md:519]: Not sure this should be changed, now it doesn't match the gif? Hmm.

File: site/en/docs/web-platform/navigation-api/index.md [hunk 13]
```diff
@@ -319,15 +541,20 @@ navigation.addEventListener('navigate', navigateEvent => {
...
-    const submitToServer = async () => {
-      await fetch(navigateEvent.destination.url, {
-        method: 'POST',
-        body: navigateEvent.formData,
-      });
-      // You could navigate again with {history: 'replace'} to change the URL here,
-      // which might indicate "done"
-    };
-    navigateEvent.transitionWhile(submitToServer());
+    navigateEvent.intercept({
```
Comment (domenic) [inline site/en/docs/web-platform/navigation-api/index.md:544]: There's a leftover canTransition above this.

**Review by domenic — COMMENTED** [review]
Thanks very much!

File: site/en/docs/web-platform/navigation-api/index.md (hunk 1 above)
Comment (jakearchibald) [reply site/en/docs/web-platform/navigation-api/index.md:47]: hah, yeah. I was passing it into `shouldNotIntercept` in an earlier draft.

File: site/en/docs/web-platform/navigation-api/index.md (hunk 4 above)
Comment (jakearchibald) [reply site/en/docs/web-platform/navigation-api/index.md:111]: I'll make them all use code font

File: site/en/docs/web-platform/navigation-api/index.md (hunk 6 above)
Comment (jakearchibald) [reply site/en/docs/web-platform/navigation-api/index.md:141]: Shall do. Although, I wonder how many folks will do this vs handle the form as they do today, then perform a replacement navigation when the POST is complete.

File: site/en/docs/web-platform/navigation-api/index.md (hunk 11 above)
Comment (jakearchibald) [reply site/en/docs/web-platform/navigation-api/index.md:452]: I'll work it into a new section but with caveats

File: site/en/docs/web-platform/navigation-api/index.md (hunk 12 above)
Comment (jakearchibald) [reply site/en/docs/web-platform/navigation-api/index.md:519]: oof, good catch, I pulled the wrong URL from my clipboard. It's supposed to be the same demo but updated for the latest API https://chief-low-specialist.glitch.me/. I'll update it.

**Comment by jakearchibald** [comment]
@rachelandrew this is ready for review and landing. The CLA check is failing due to @domenic, but he's a Googler so I presume it's fine?

**Comment by jakearchibald** [comment]
Is there anything I need to do before this can be reviewed and merged?

**Comment by rachelandrew** [comment]
@jakearchibald no - I just missed the original message (feel free to ping me on Chat if I'm ignoring something). I'll take a look now.

File: site/en/docs/web-platform/navigation-api/index.md [hunk 14]
```diff
@@ -33,47 +34,62 @@ If you'd like to read the technical proposal, [check out the Draft Report in the
...
+- Calling `intercept({ handler })` (as described above) to handle the navigation.
 - Calling `preventDefault()`, which can cancel the navigation completely.
 
-This example calls `transitionWhile()` on the event with a promise generated from async functions.
-By calling this method, the browser knows that your code will configure the next state of your site.
-This will create a transition object, `navigation.transition`, which other code can use to track the progress of the transition.
+{% Aside %}
+An earlier version of this API used `navigateEvent.transitionWhile(promise)` rather than `navigateEvent.intercept({ handler })`.
+
+`intercept` landed in Chrome 105.
```
Comment (rachelandrew) [inline site/en/docs/web-platform/navigation-api/index.md:69]:
```suggestion
`intercept` is available from Chrome 105.
```

File: site/en/docs/web-platform/navigation-api/index.md [hunk 15]
```diff
@@ -33,47 +34,62 @@ If you'd like to read the technical proposal, [check out the Draft Report in the
...
+`transitionWhile` is deprecated, and will be removed in Chrome 108.
+{% endAside %}
+
+This example calls `intecept()` on the event.
```
Comment (rachelandrew) [inline site/en/docs/web-platform/navigation-api/index.md:73]:
```suggestion
This example calls `intercept()` on the event.
```

File: site/en/docs/web-platform/navigation-api/index.md [hunk 16]
```diff
@@ -92,90 +108,253 @@ Links might come and go on your page, and they're not the only way users can nav
 E.g., they may submit a form or even use an [image map].
 Your page might deal with these, but there's a long tail of possibilities which could just be simplified—something that the new Navigation API achieves.
 
+Additionally, the above doesn't handle back/forward navigation. There's another event for that, `"popstate"`.
+
 Personally, the History API often _feels_ like it could go some way to help with these possibilities.
 However, it really only has two surface areas: responding if the user presses Back or Forward in their browser, plus pushing and replacing URLs.
-It doesn't have an analogy to "navigate", except if you manually set up listeners for, e.g., click events, as demonstrated above.
+It doesn't have an analogy to `"navigate"`, except if you manually set up listeners for, e.g., click events, as demonstrated above.
```
Comment (rachelandrew) [inline site/en/docs/web-platform/navigation-api/index.md:115]:
```suggestion
It doesn't have an analogy to `"navigate"`, except if you manually set up listeners for click events for example, as demonstrated above.
```

Comment (rachelandrew) [reply site/en/docs/web-platform/navigation-api/index.md:115]: docs style guide suggests not using e.g. or i.e.

File: site/en/docs/web-platform/navigation-api/index.md [hunk 17]
```diff
@@ -92,90 +108,253 @@ Links might come and go on your page, and they're not the only way users can nav
...
+  In this case the old navigation is abandoned in favour of the new navigation.
+- The user clicks the 'stop' button in the browser.
 
-To deal with any of these possibilities, the event passed to the "navigate" listener contains a `signal` property, which is an `AbortSignal`.
+To deal with any of these possibilities, the event passed to the `"navigate"` listener contains a `signal` property, which is an `AbortSignal`.
 For more information see [Abortable fetch][abortable-fetch].
+
 The short version is it basically provides an object that fires an event when you should stop your work.
 Notably, you can pass an `AbortSignal` to any calls you make to `fetch()`, which will cancel in-flight network requests if the navigation is preempted.
 This will both save the user's bandwidth, and reject the `Promise` returned by `fetch()`, preventing any following code from e.g., updating the DOM to show a now invalid page navigation.
```
Comment (rachelandrew) [inline site/en/docs/web-platform/navigation-api/index.md:237]:
```suggestion
This will both save the user's bandwidth, and reject the `Promise` returned by `fetch()`, preventing any following code from actions such as updating the DOM to show a now invalid page navigation.
```

**Comment by Miami87** [comment]
Rachel Andrew ***@***.***> schrieb am Do., 18. Aug. 2022,
10:58:

**Review by rachelandrew — APPROVED** [review]
Couple of small suggestions and a typo. I'll sort out the CLA override

File: site/en/docs/web-platform/navigation-api/index.md (hunk 15 above)
Comment (jakearchibald) [reply site/en/docs/web-platform/navigation-api/index.md:73]: oh wow, good catch

**Comment by jakearchibald** [comment]
> I just missed the original message (feel free to ping me on Chat if I'm ignoring something). I'll take a look now.

Shall do! Didn't want to hassle you. Also, should I be pinging you directly, or is there a group I can send this stuff to?

**Comment by rachelandrew** [comment]
You can always ask in the dc.c. chat if it's just something needing a random tech writer LGTM, but if you've directly assigned something to me and I'm ignoring it (or it's urgent) just prod me on chat.
